File: backend/src/utils/files/uploadPaths.ts
import fs from "fs";
import path from "path";
import { AppError } from "../../lib/errors/AppError";

export const UPLOADS_ROOT = process.env.UPLOAD_PATH
  ? path.resolve(process.env.UPLOAD_PATH)
  : path.resolve(__dirname, "../../../uploads");

/** Resolves a user-provided filename to a path under uploads (no directory traversal). */
export function resolveUploadPath(filename: string): string {
  if (!filename || typeof filename !== "string") {
    throw new AppError("File reference is required", 400);
  }

  const base = path.basename(filename);
  const resolved = path.resolve(UPLOADS_ROOT, base);

  if (!resolved.startsWith(UPLOADS_ROOT + path.sep) && resolved !== UPLOADS_ROOT) {
    throw new AppError("Invalid file path", 400);
  }

  return resolved;
}

export function assertUploadExists(filePath: string, notFoundMessage: string): void {
  if (!fs.existsSync(filePath)) {
    throw new AppError(notFoundMessage, 404);
  }
}

export default resolveUploadPath;

File: backend/app/utils/upload_paths.py
from pathlib import Path

from app.config import UPLOAD_PATH
from app.errors import AppError


def resolve_upload_path(filename: str) -> Path:
    if not filename or not filename.strip():
        raise AppError("Filename is required", 400)
    safe_name = Path(filename).name
    if not safe_name:
        raise AppError("Filename is required", 400)
    return UPLOAD_PATH / safe_name


def assert_upload_exists(filename: str) -> Path:
    path = resolve_upload_path(filename)
    if not path.is_file():
        raise AppError("File not found", 404)
    return path
